File: vital-core/src/main/java/me/vitalframework/Vital.java
package me.vitalframework;

import lombok.Getter;
import lombok.NonNull;
import lombok.SneakyThrows;
import org.springframework.boot.Banner;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.core.env.Environment;
import org.springframework.core.env.StandardEnvironment;
import org.springframework.core.io.DefaultResourceLoader;

import java.io.PrintStream;
import java.util.Properties;

/**
 * The main instance of the Vital-Framework.
 *
 * @author xRa1ny
 */
public class Vital {
    @Getter
    static ConfigurableApplicationContext context;

    /**
     * Initializes the Vital-Framework using Spring Boot
     *
     * @param plugin The plugin instance itself
     */
    @SneakyThrows
    public static <T> void run(@NonNull T plugin) {
        final var pluginClassLoader = plugin.getClass().getClassLoader();

        // needed or else spring startup fails
        Thread.currentThread().setContextClassLoader(pluginClassLoader);

        final var loader = new DefaultResourceLoader(pluginClassLoader);
        final var builder = new SpringApplicationBuilder();
        final var pluginConfiguration = Class.forName(plugin.getClass().getPackageName() + ".PluginConfiguration");

        try {
            final var properties = new Properties();

            properties.load(pluginClassLoader.getResourceAsStream("application.properties"));

            properties.forEach((key, value) -> System.setProperty(key.toString(), value.toString()));
        } catch (Exception ignored) {
            // if we haven't defined an application.properties file, we may skip this step
        }

        context = builder.sources(pluginConfiguration)
                .initializers(applicationContext -> {
                    applicationContext.getBeanFactory().registerSingleton("plugin", plugin);
                    applicationContext.setClassLoader(pluginClassLoader);
                    applicationContext.setEnvironment(new StandardEnvironment());
                })
                .resourceLoader(loader)
                .banner(new VitalBanner())
                .logStartupInfo(false)
                .run();
    }

    public static class VitalBanner implements Banner {
        @Override
        public void printBanner(Environment environment, Class<?> sourceClass, PrintStream out) {
            out.print("""
                      .
                     /\\\\ __     ___ _        _  ______ \s
                    ( ( )\\ \\   / ( ) |_ __ _| | \\ \\ \\ \\\s
                     \\\\/  \\ \\ / /| | __/ _` | |  \\ \\ \\ \\
                      ,    \\ V / | | || (_| | |  / / / /
                    ========\\_/==|_|\\__\\__,_|_|=/_/_/_/\s
                                                       \s
                    """);
        }
    }
}

File: vital-core/src/main/kotlin/me/vitalframework/Vital.kt
package me.vitalframework

import me.vitalframework.VitalCoreSubModule.Companion.logger
import org.springframework.boot.SpringApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.Configuration
import org.springframework.util.ClassUtils
import java.util.Properties

object Vital {
    private val logger = logger()

    @JvmStatic
    lateinit var context: ConfigurableApplicationContext
        private set

    @JvmStatic
    val vitalSubModules = mutableListOf<String>()

    @JvmStatic
    lateinit var metadata: Metadata
        private set

    @JvmStatic
    fun run(
        loader: Any,
        classLoader: ClassLoader,
    ) {
        logger.debug("Running Vital via plugin loader '{}' and class loader '{}'...", loader, classLoader)

        Thread.currentThread().contextClassLoader = classLoader
        ClassUtils.overrideThreadContextClassLoader(classLoader)

        logger.debug("Vital class loader successfully overridden to '{}'", classLoader)

        // load metadata from "vital-metadata.properties"
        loadMetadata(classLoader)

        logger.debug("Loading main class '${metadata.mainClassName}'...")
        val mainClass = Class.forName(metadata.mainClassName)
        logger.debug("Main class '${metadata.mainClassName}' successfully loaded")

        // start up spring boot using the previously generated "PluginConfiguration" class as the main class
        logger.debug("Running spring boot...")
        context =
            SpringApplicationBuilder(classLoader.loadClass("${mainClass.packageName}.PluginConfiguration"))
                // here we register the plugin instance as a bean so we can inject it elsewhere
                .initializers({ it.beanFactory.registerSingleton("plugin", loader) })
                // this is needed so spring can locate classes and resources that are on the plugin classpath
                .resourceLoader(VitalResourceLoader())
                .run()
    }

    @JvmStatic
    fun exit() {
        logger.info("Shutting down Vital...")
        val exitCode = SpringApplication.exit(context)
        logger.info("Vital exited with code '$exitCode'")
    }

    private fun loadMetadata(classLoader: ClassLoader) {
        logger.debug("Loading Vital metadata...")

        val metadataProperties = Properties().apply { load(classLoader.getResourceAsStream(Metadata.FILE_NAME)) }
        metadata = Metadata(metadataProperties[Metadata.Property.MAIN_CLASS].toString())

        logger.debug("Vital metadata loaded successfully, {}", metadata)
    }

    data class Metadata(
        val mainClassName: String,
    ) {
        val mainClass: Class<*> get() = Class.forName(mainClassName)

        /**
         * Serializes this Metadata instance into a properties file string.
         */
        fun serialize() =
            """
            ${Property.MAIN_CLASS}=$mainClassName
            """.trimIndent()

        companion object {
            const val FILE_NAME = "vital-metadata.properties"
        }

        object Property {
            const val MAIN_CLASS = "vital.main-class"
        }
    }

    @Configuration
    @Target(AnnotationTarget.CLASS)
    @Retention(AnnotationRetention.RUNTIME)
    annotation class Info(
        val name: String,
        val description: String,
        val apiVersion: String,
        val version: String,
        val author: Array<String>,
        val environment: PluginEnvironment,
        /**
         * This instructs Vital to scan additional packages for Vital functionality.
         * If you have to include an external dependency in your plugin that uses Vital,
         * like Commands, Configs, etc. this will be the place to define those packages
         */
        val scanAdditionalPackages: Array<String> = [],
    ) {
        companion object {
            val DEFAULT_PACKAGES = arrayOf("me.vitalframework")
        }

        enum class PluginEnvironment(
            val ymlFileName: String,
        ) {
            SPIGOT("plugin.yml"),
            PAPER("plugin.yml"),
            BUNGEE("bungee.yml"),
        }
    }
}
